classes/controlledVocab/ControlledVocab.php

```diff
@@ -1,108 +1,201 @@
+            set: fn($value) => [$this->primaryKey => $value],
+        )->shouldCache();
     }
 
     /**
-     * Get associated type.
-     *
-     * @return int
+     * Get the list of pre defined vocab symbolics
      */
-    public function getAssocType()
+    public static function getDefinedVocabSymbolic(): array
     {
-        return $this->getData('assocType');
+        return [
+            static::CONTROLLED_VOCAB_SUBMISSION_AGENCY,
+            static::CONTROLLED_VOCAB_SUBMISSION_DISCIPLINE,
+            static::CONTROLLED_VOCAB_SUBMISSION_KEYWORD,
+            static::CONTROLLED_VOCAB_SUBMISSION_LANGUAGE,
+            static::CONTROLLED_VOCAB_SUBMISSION_SUBJECT,
+        ];
     }
 
     /**
-     * Set associated type.
-     *
-     * @param int $assocType
+     * Check if a provided vocab symbolic defined in pre defined symbolic list
      */
-    public function setAssocType($assocType)
+    public static function hasDefinedVocabSymbolic(string $vocab): bool
     {
-        $this->setData('assocType', $assocType);
+        return in_array($vocab, static::getDefinedVocabSymbolic());
     }
 
     /**
-     * Get symbolic name.
-     *
-     * @return string
+     * Get all controlled vocab entries for this controlled vocab
      */
-    public function getSymbolic()
+    public function controlledVocabEntries(): HasMany
     {
-        return $this->getData('symbolic');
+        return $this->hasMany(ControlledVocabEntry::class, 'controlled_vocab_id', 'controlled_vocab_id');
     }
 
     /**
-     * Set symbolic name.
-     *
-     * @param string $symbolic
+     * Scope a query to only include vocabs with a specific symbolic.
+     */
+    public function scopeWithSymbolic(Builder $query, string $symbolic): Builder
+    {
+        return $query->where('symbolic', $symbolic);
+    }
+
+    /**
+     * Scope a query to only include vocabs with a specific assoc type and assoc ID.
      */
-    public function setSymbolic($symbolic)
+    public function scopeWithAssoc(Builder $query, int $assocType, int $assocId): Builder
     {
-        $this->setData('symbolic', $symbolic);
+        return $query
+            ->where('assoc_type', $assocType)
+            ->where('assoc_id', $assocId);
+    }
+
+    /**
+     * Scope a query to only include vocabs associated with given context id
+     */
+    public function scopeWithContextId(Builder $query, int $contextId): Builder
+    {
+        return $query
+            ->where(
+                fn ($query) => $query
+                    ->select('context_id')
+                    ->from('submissions')
+                    ->whereColumn(
+                        DB::raw(
+                            "(SELECT publications.submission_id 
+                            FROM publications 
+                            INNER JOIN {$this->table} 
+                            ON publications.publication_id = {$this->table}.assoc_id 
```

Comment:
@asmecher I initially thought there can be a easier alternative but it will be bit extra work . First of all, this is how we were retrieving all the vocabs for a given context and the same concept (from controlled_vocabs -> publications -> submissions to get the context id ) used previous code base , just updated newer query builder mechanism . 

However if I am correct, we would like to allow different types of assoc , not just the publication one as we are doing it now . in that case, as still sometimes need to figure out vocabs for a given context id , it will become too difficult if there are multiple type assoc get associated with for a multi context system . 

My proposal is to allow the `context_id` column to `controlled_vocabs` table and that will solve this case with a lot easier approach .

Reply:
I think the only place this function is used is in the vocabs API controller, which is where the front end gets autosuggest possibilities. I think having a single endpoint for all vocab suggestions is actually the mistake -- it leads to the need for a single point of access for vocabs that might be attached to many different things, so we have to hard-code all the checks to make sure the list is allowed into a single place.

Instead, I suggest we break the endpoints where autosuggest responses are built into the appropriate places in the API. Autosuggestions for reviewer interests, which are global, should not be attached to the same endpoint as autosuggest results for article keywords, which are (probably) journal-specific.

@ewhanson, can we revisit the controlled vocab API handler? I don't think a single endpoint is the right design.